nistats/reporting/glm_reporter.py

```diff
@@ -0,0 +1,561 @@
+        
+    plot_type: String. ['stat' (default)| 'glass']
+        Specifies the type of plot to be drawn for the statistical maps.
+        
+    Returns
+    -------
+    report_text: HTMLDocument Object
+        Contains the HTML code for the GLM Report.
+    """
+    if not display_mode:
+        if plot_type == 'stat':
+            display_mode = 'z'
+        elif plot_type == 'glass':
+            display_mode = 'lzry'
+    cluster_threshold = cluster_threshold if cluster_threshold else 0
+    pd.set_option('display.max_colwidth', -1)
+    bg_img = load_mni152_template() if bg_img == 'MNI 152 Template' else bg_img
+    html_template_path = os.path.join(html_template_root_path,
+                                      'report_template.html')
+    with open(html_template_path) as html_file_obj:
+        html_template_text = html_file_obj.read()
+    report_template = string.Template(html_template_text)
+    
+    contrasts = _make_contrasts_dict(contrasts)
+    contrast_plots = _make_dict_of_contrast_plots(contrasts, model)
+    all_contrast_plots_html = ['{}<p>{}'.format(key, item)
+                           for key, item in contrast_plots.items()
+                           ]
+    all_contrast_plots_text = '<p>'.join(all_contrast_plots_html)
+
+    page_title, page_heading_1, page_heading_2 = _make_page_title_heading(
+        contrasts,
+        title,
+        )
+    pd.set_option('display.max_colwidth', 50)
+    model_attributes_html = _make_model_attributes_html_table(model)
+    statistical_maps = make_statistical_maps(model, contrasts)
+    html_design_matrices = _make_html_for_design_matrices(model)
+    roi_plot_html_code = _make_roi_plot(roi_img, bg_img)
+    all_components = _make_report_components(
+            statistical_maps=statistical_maps,
+            contrasts_plots=contrast_plots,
+            threshold=threshold,
+            alpha=alpha,
+            cluster_threshold=cluster_threshold,
+            height_control=height_control,
+            min_distance=min_distance,
+            bg_img=bg_img,
+            display_mode=display_mode,
+            plot_type=plot_type,
+            )
+    all_components_text = '\n'.join(all_components)
+    report_values = {'page_title': page_title,
+                     'page_heading_1': page_heading_1,
+                     'page_heading_2': page_heading_2,
+                     'model_attributes': model_attributes_html,
+                     'all_contrasts_with_plots': all_contrast_plots_text,
+                     'design_matrices': html_design_matrices,
+                     'roi_plot': roi_plot_html_code,
+                     'component': all_components_text,
+                     }
+    report_text = report_template.safe_substitute(**report_values)
+    report = HTMLDocument(report_text)
+    report.width = nb_width  # for better visual experience in Jupyter Notebooks.
+    report.height = nb_height
+    return report
+
+
+def _make_contrasts_dict(contrasts):
```

Comment:
Necessary to write unit tests for this.

We have several ways of specifying contrasts, & it's important we test that they are all transformed correctly & to the same format for the reporting code to not break.

Reply:
note that this will raise a TypeError (unhashable type) if contrasts is a list of arrays. 

> We have several ways of specifying contrasts, & it's important we test that they are all transformed correctly & to the same format for the reporting code to not break.

there should be a description of these ways in the docstring and a check that validates the provided contrasts

Reply:
> check that validates the provided contrasts

The `compute_contrasts` method of FLM & SLM would do that. No sense in me adding more code for the same task.